Allow to specifically unset the manufacturer id in zha

jayme-github opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
I don't think 0 is a valid manufacturer id so I suppose this is not a breaking change.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This change allows to disable sending the manufacturer id (by providing 0 as manufacturer_code override.

Some devices (Sonoff TRVZB) don't accept reading/writing attributes in the manufacturer specific cluster when the manufacturer id is set.

I'm not sure if this is a desirable change at all (I'm not familiar with developing ZigBee stuff or the home-assistant codebase) but it helped me debug/control some attributes while playing with the mentioned TRV.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

jayme-github committed:
Allow to specifically unset the manufacturer id

This change allows to disable sending the manufacturer id (by providing
0 as manufacturer_code override.

Some devices (Sonoff TRVZB) don't accept reading/writing attributes in
the manufacturer specific cluster when the manufacturer id is set.

elupus reviewed (commented):
Im thinking we should have a sentinel default value for the manufacturer, the sentinel is what trigger the use of the standard manufacturer. Then allow an explict call with None as manufacturer instead.

TheJulianJES commented:
> Some devices (Sonoff TRVZB) don't accept reading/writing attributes in the manufacturer specific cluster when the manufacturer id is set.

So there's already a quirk to expose the manufacturer specific attributes?
Shouldn't the quirk just be changed to not send the manufacturer id for those attributes then (so they're added not added as manufacturer specific attributes).

dmulcahey commented:
I agree with @TheJulianJES. IIRC this is already handled and the attribute definitions should change.

jayme-github commented:
Thanks all for taking the time looking at this and responding.
I indeed failed to mention that I was toying around with writing a quirk for that device when hitting this (sorry). I did override `_is_manuf_specific` in my CustomCluster to always return False but that did not seem to have an impact when manually reading/writing attributes, though. But it is very possible that I'm missing something here as I'm very new to this topic and it seems to be quite a learning curve zha/zha-quirks wise. ;)

TheJulianJES commented:
Can you create an issue for that device in the zha-device-handlers repo, provide your quirk, and tag me?

jayme-github committed:
Do not exlicitely override the manufacturer id

Let zigpy take care of adding the manufacturer id in case it is
required (e.g. read/write from a manufacturer specific cluster).

jayme-github commented:
I'm not sure how to proceed here. The last commit hands the decision to set/specify the manufacturer id down to zigpy (IIUC) if the user does not specify a value in the dialog. This works for me with manufacturer specific clusters that require not to set manufacturer id (as in the linked quirk).

SuperSandro2000 commented:
> Can you create an issue for that device in the zha-device-handlers repo, provide your quirk, and tag me?

There is already https://github.com/zigpy/zha-device-handlers/issues/2772 and https://github.com/zigpy/zha-device-handlers/pull/2831

elupus commented:
I came to the exact same conclusion here: https://github.com/home-assistant/core/pull/105453 this code has no business setting manufacturer id. We don't do that when attributes are used by entities (numbers, switches, ...) so it make no sense forcing it here.

I took it a bit further in my change to allow the forced off value too.

dmulcahey commented:
Sorry for the confusion. Superseded by https://github.com/home-assistant/core/pull/108257
